src/modules/cache.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2016 The Noise-rs Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use num_traits::Float;
+use math::{Point2, Point3, Point4};
+use NoiseModule;
+use std::cell::RefCell;
+
+/// Noise module that caches the last output value generated by the source
+/// module.
+///
+/// If an application passes input coordinates to the `get()` method that
+/// differs from the previously passed-in value, this noise module calls
+/// the source module to calculate the output value. The output value, as well
+/// as the input coordinates, are stored (cached) in this noise module.
+///
+/// If the application passes input coordinates to the `get()` method that is
+/// equal to the previously passed-in input coordinates, this noise module
+/// returns the cached output value without having the source module
+/// recalculate the output value.
+///
+/// Caching a noise module is useful if it is used as a source module for
+/// multiple noise modules. If a source module is not cached, the source
+/// module will redundantly calculate the same output value once for each
+/// noise module in which it is included.
+#[derive(Clone, Debug)]
+pub struct Cache<Source, T> {
+    /// Outputs the value to be cached.
+    pub source: Source,
+
+    value: RefCell<Option<T>>,
+
+    point: RefCell<Vec<T>>,
```

Comment:
If I added a type parameter for the point `P` so that it had

`pub struct Cache<Source, T, P> {
point: RefCell<P>`

then on instantiation of Cache, it would have the type signature Cache::<Source, T, Point3<T>> for a 3D point, and be locked into that signature on compilation. This would prevent re-using the Cache module for a 2D or 4D point without creating another instance of Cache with the correct type signature. This type signature lock-in would propagate through the entire module chain that contains Cache, since they would not be able to feed or retrieve a non-3D point through Cache.

The point of the modules I'm making here is to make them as dimensionally-agnostic as possible, so that whatever module chain is created, can pass through any of the implemented N-dimensional points without having to specify the dimensionality of the point for each module.

As it stands, I can create a module chain, then feed a set of 2D points through it, and then re-use that chain and feed 3D points through it, without re-implementing any modules or changing any definitions. Adding P to Cache would prevent this genericness.